subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/execution/ProgramParser.kt

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.execution
+
+import org.gradle.kotlin.dsl.support.compilerMessageFor
+
+import org.jetbrains.kotlin.lexer.KotlinLexer
+import org.jetbrains.kotlin.lexer.KtTokens
+
+
+internal
+object ProgramParser {
+
+    fun parse(source: ProgramSource, kind: ProgramKind, target: ProgramTarget): Program = try {
+        programFor(source, kind, target)
+    } catch (unexpectedBlock: UnexpectedBlock) {
+        handleUnexpectedBlock(unexpectedBlock, source.text, source.path)
+    }
+
+    private
+    fun programFor(source: ProgramSource, kind: ProgramKind, target: ProgramTarget): Program {
+
+        val sourceWithoutComments =
+            source.map { it.erase(commentsOf(it.text)) }
+
+        val buildscriptFragment =
+            topLevelFragmentFrom(sourceWithoutComments, if (target == ProgramTarget.Gradle) "initscript" else "buildscript")
+
+        val pluginsFragment =
+            if (kind == ProgramKind.TopLevel && target == ProgramTarget.Project) topLevelFragmentFrom(sourceWithoutComments, "plugins")
+            else null
+
+        val buildscript =
+            buildscriptFragment?.takeIf { it.isNotBlank() }?.let(Program::Buildscript)
+
+        val plugins =
+            pluginsFragment?.takeIf { it.isNotBlank() }?.let(Program::Plugins)
+
+        val stage1 =
+            buildscript?.let { bs ->
+                plugins?.let { ps ->
+                    Program.Stage1Sequence(bs, ps)
+                } ?: bs
+            } ?: plugins
+
+        val remainingSource =
+            sourceWithoutComments.map {
+                it.erase(
+                    listOfNotNull(
+                        buildscriptFragment?.section?.wholeRange,
+                        pluginsFragment?.section?.wholeRange))
+            }
+
+        val stage2 = remainingSource
+            .takeIf { it.text.isNotBlank() }
+            ?.let(Program::Script)
+
+        stage1?.let { s1 ->
+            return stage2?.let { s2 ->
+                Program.Staged(s1, s2)
+            } ?: s1
+        }
+
+        return stage2
+            ?: Program.Empty
+    }
+
+    private
+    fun topLevelFragmentFrom(source: ProgramSource, identifier: String): ProgramSourceFragment? =
+        extractTopLevelBlock(source.text, identifier)
+            ?.let { source.fragment(it) }
+
+    private
+    fun ProgramSourceFragment.isNotBlank() =
+        source.text.subSequence(section.block.start + 1, section.block.endInclusive).isNotBlank()
+}
+
+
+private
+fun commentsOf(script: String): List<IntRange> =
+    KotlinLexer().run {
+        val comments = mutableListOf<IntRange>()
+        start(script)
+        while (tokenType != null) {
+            if (tokenType in KtTokens.COMMENTS) {
+                comments.add(tokenStart..(tokenEnd - 1))
+            }
+            advance()
+        }
+        comments
+    }
```

Comment:
Is the `IntRange` the lines that the block comments are on? Does this handle non-block comments?

Reply:
Yes, see `ProgramParserTest`.